kubectl panic: runtime error: invalid memory address or nil pointer dereference

I am getting this runtime error on kubectl create:

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.7+e4e6878", GitCommit:"e4e6878293a339e4087dae684647c9e53f1cf9f0", GitTreeState:"not a git tree"}
Server Version: version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.3", GitCommit:"6a81b50c7e97bbe0ade075de55ab4fa34f049dc2", GitTreeState:"clean"}

$ kubectl --server=http://localhost:18080 create -f test-rc.yaml
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0xa0 pc=0x3b5cdc]

goroutine 1 [running]:
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x0, 0x0, 0xc8205e7c80, 0x28, 0xc82056a130, 0x6, 0x0, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:241 +0x127c
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x74c480, 0xc8203d4900, 0xc8205e7c20, 0x25, 0xc82056a150, 0x5, 0xc8205d9680, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:258 +0xc87
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateObject(0xc8202f80c0, 0x7656e0, 0xc8205e7890, 0xc8205e7bf0, 0x21, 0xc8204400a0, 0xc, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:208 +0x11d0
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x7656e0, 0xc8205e7890, 0xc8203d4de0, 0x20, 0xc8204400a0, 0xc, 0x0, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:231 +0xd3
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x74c480, 0xc8203d49c0, 0xc8203d4dc0, 0x1d, 0xc8204400c0, 0x5, 0xc8205d95f0, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:258 +0xc87
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateObject(0xc8202f80c0, 0x7656e0, 0xc8205e7860, 0xc8203d4ba0, 0x13, 0xc82025bbe0, 0xa, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:208 +0x11d0
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x7656e0, 0xc8205e7860, 0xc8203d4b40, 0x12, 0xc82025bbe0, 0xa, 0xc8205d9200, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:231 +0xd3
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateObject(0xc8202f80c0, 0x7656e0, 0xc8205e77d0, 0xc82056b440, 0xe, 0xc82040fbe0, 0x12, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:208 +0x11d0
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x7656e0, 0xc8205e77d0, 0xc82056b430, 0xd, 0xc82040fbe0, 0x12, 0xc8205d9050, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:231 +0xd3
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateObject(0xc8202f80c0, 0x7656e0, 0xc8205e7770, 0xc82056b400, 0x5, 0xc82000e180, 0x1c, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:208 +0x11d0
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).validateField(0xc8202f80c0, 0x7656e0, 0xc8205e7770, 0xc82056ae80, 0x4, 0xc82000e180, 0x1c, 0xc8205d8d80, 0x0, 0x0, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:231 +0xd3
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateObject(0xc8202f80c0, 0x7656e0, 0xc8205e76e0, 0x0, 0x0, 0xc8203d4a40, 0x18, 0x0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:208 +0x11d0
k8s.io/kubernetes/pkg/api/validation.(*SwaggerSchema).ValidateBytes(0xc8202f80c0, 0xc820302000, 0x2ff, 0x300, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/schema.go:156 +0x8a1
k8s.io/kubernetes/pkg/kubectl/cmd/util.getSchemaAndValidate(0x13a03d0, 0xc8202660a0, 0xc820302000, 0x2ff, 0x300, 0xa08c30, 0x3, 0xc8203005e0, 0x2, 0xc8200f2520, ...)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go:357 +0x8a1
k8s.io/kubernetes/pkg/kubectl/cmd/util.(*clientSwaggerSchema).ValidateBytes(0xc8202569f0, 0xc820302000, 0x2ff, 0x300, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory.go:379 +0x803
k8s.io/kubernetes/pkg/kubectl/resource.ValidateSchema(0xc820302000, 0x2ff, 0x300, 0x13a0178, 0xc8202569f0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:210 +0x1e4
k8s.io/kubernetes/pkg/kubectl/resource.(*StreamVisitor).Visit(0xc8202e2c00, 0xc8202f36a0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:481 +0x2fc
k8s.io/kubernetes/pkg/kubectl/resource.(*FileVisitor).Visit(0xc8202f35a0, 0xc8202f36a0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:441 +0x12d
k8s.io/kubernetes/pkg/kubectl/resource.EagerVisitorList.Visit(0xc8202e8e70, 0x1, 0x1, 0xc8202ea990, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:186 +0x12c
k8s.io/kubernetes/pkg/kubectl/resource.(*EagerVisitorList).Visit(0xc8202f35c0, 0xc8202ea990, 0x0, 0x0)
	<autogenerated>:98 +0xb7
k8s.io/kubernetes/pkg/kubectl/resource.FlattenListVisitor.Visit(0x13a02b8, 0xc8202f35c0, 0xc8202ea930, 0xc8202e2c80, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:361 +0xaa
k8s.io/kubernetes/pkg/kubectl/resource.(*FlattenListVisitor).Visit(0xc8202f35e0, 0xc8202e2c80, 0x0, 0x0)
	<autogenerated>:111 +0xb7
k8s.io/kubernetes/pkg/kubectl/resource.DecoratedVisitor.Visit(0x13a02e0, 0xc8202f35e0, 0xc8202f3620, 0x3, 0x4, 0xc8202f3660, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:276 +0xf0
k8s.io/kubernetes/pkg/kubectl/resource.(*DecoratedVisitor).Visit(0xc8202ea960, 0xc8202f3660, 0x0, 0x0)
	<autogenerated>:127 +0xac
k8s.io/kubernetes/pkg/kubectl/resource.ContinueOnErrorVisitor.Visit(0x13a0308, 0xc8202ea960, 0xc8202e2c40, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/visitor.go:294 +0x12f
k8s.io/kubernetes/pkg/kubectl/resource.(*ContinueOnErrorVisitor).Visit(0xc8202e8e90, 0xc8202e2c40, 0x0, 0x0)
	<autogenerated>:123 +0xae
k8s.io/kubernetes/pkg/kubectl/resource.(*Result).Visit(0xc8202f4230, 0xc8202e2c40, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/resource/result.go:79 +0x6d
k8s.io/kubernetes/pkg/kubectl/cmd.RunCreate(0xc8201174a0, 0xc82000d180, 0x13643d0, 0xc820030010, 0xc8201d11c0, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/create.go:128 +0x547
k8s.io/kubernetes/pkg/kubectl/cmd.NewCmdCreate.func1(0xc82000d180, 0xc8201fed80, 0x0, 0x3)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/create.go:61 +0xf8
github.com/spf13/cobra.(*Command).execute(0xc82000d180, 0xc8201fecf0, 0x3, 0x3, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/Godeps/_workspace/src/github.com/spf13/cobra/command.go:499 +0x869
github.com/spf13/cobra.(*Command).Execute(0xc82000cc40, 0x0, 0x0)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/Godeps/_workspace/src/github.com/spf13/cobra/command.go:582 +0x46a
main.main()
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:30 +0xf9

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/Cellar/go/1.5.3/libexec/src/runtime/asm_amd64.s:1721 +0x1

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0xedf360)
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x67
created by github.com/golang/glog.init.1
	/private/tmp/kubernetes-cli20160129-22889-8puto1/kubernetes-1.1.7/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x297

goroutine 10 [syscall]:
os/signal.loop()
	/usr/local/Cellar/go/1.5.3/libexec/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/Cellar/go/1.5.3/libexec/src/os/signal/signal_unix.go:28 +0x37

goroutine 14 [IO wait]:
net.runtime_pollWait(0x136c228, 0x72, 0xc820012160)
	/usr/local/Cellar/go/1.5.3/libexec/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc82022ad10, 0x72, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc82022ad10, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc82022acb0, 0xc820236000, 0x1000, 0x1000, 0x0, 0x1360050, 0xc820012160)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8200303f8, 0xc820236000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x136c2e8, 0xc8200303f8, 0xc8202007e8, 0xc820236000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc820232a00, 0xc820236000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc820234240)
	/usr/local/Cellar/go/1.5.3/libexec/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc820234240, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.5.3/libexec/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc820200790)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/Cellar/go/1.5.3/libexec/src/net/http/transport.go:685 +0xc78

goroutine 15 [select]:
net/http.(*persistConn).writeLoop(0xc820200790)
	/usr/local/Cellar/go/1.5.3/libexec/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/Cellar/go/1.5.3/libexec/src/net/http/transport.go:686 +0xc9d
```

And this is test-rc.yaml file:

```
apiVersion: v1
kind: ReplicationController
metadata:
  name: kube-dns-v10
  namespace: kube-system
  labels:
    k8s-app: kube-dns
    version: v10
    kubernetes.io/cluster-service: "true"
spec:
  replicas: 1
  selector:
    k8s-app: kube-dns
    version: v10
  template:
    metadata:
      labels:
        k8s-app: kube-dns
        version: v10
        kubernetes.io/cluster-service: "true"
    spec:
      containers:
      - name: carbon-relay
        image: banno/carbon-relay
        resources:
          limits:
            cpu: 100m
            memory: 50Mi
          requests:
            cpu: 100m
            memory: 50Mi
        env:
        - name: DESTINATIONS
          # Azure DNS
          value: 192.168.81.25:17124
        args:
        -
      volumes:
      - name: etcd-storage
        emptyDir: {}
      - hostPath:
          path: /etc/ssl/certs
        name: ssl-certs-host
      dnsPolicy: Default
```

I can see that the empty `args: -` may be causing this but I'm sure we can show a better error for this.

--- a/hack/test-cmd.sh
+++ b/hack/test-cmd.sh
@@ -884,6 +884,18 @@ __EOF__
   kubectl delete hpa frontend "${kube_flags[@]}"
   kubectl delete rc  frontend "${kube_flags[@]}"
 
+  ## kubectl create should not panic on empty string lists in a template
+  ERROR_FILE="${KUBE_TEMP}/validation-error"
+  kubectl create -f hack/testdata/invalid-rc-with-empty-args.yaml "${kube_flags[@]}" 2> "${ERROR_FILE}" || true
+  # Post-condition: should get an error reporting the empty string
+  if grep -q "unexpected nil value for field" "${ERROR_FILE}"; then
+    kube::log::status "\"kubectl create with empty string list returns error as expected: $(cat ${ERROR_FILE})"
+  else
+    kube::log::status "\"kubectl create with empty string list returns unexpected error or non-error: $(cat ${ERROR_FILE})"
+    exit 1
+  fi
+  rm "${ERROR_FILE}"
+
   ## kubectl apply should create the resource that doesn't exist yet
   # Pre-Condition: no POD exists
   kube::test::get_object_assert pods "{{range.items}}{{$id_field}}:{{end}}" ''

--- a/hack/testdata/invalid-rc-with-empty-args.yaml
+++ b/hack/testdata/invalid-rc-with-empty-args.yaml
@@ -1,0 +1,20 @@
+apiVersion: v1
+kind: ReplicationController
+metadata:
+  name: kube-dns-v10
+  namespace: kube-system
+spec:
+  replicas: 1
+  selector:
+    k8s-app: kube-dns
+  template:
+    metadata:
+      labels:
+        k8s-app: kube-dns
+    spec:
+      containers:
+      - name: carbon-relay
+        image: banno/carbon-relay
+        args:
+        -
+# above is the empty arg string.

--- a/pkg/api/validation/schema.go
+++ b/pkg/api/validation/schema.go
@@ -297,6 +297,10 @@ func (s *SwaggerSchema) isGenericArray(p swagger.ModelProperty) bool {
 var versionRegexp = regexp.MustCompile(`^v.+\..*`)
 
 func (s *SwaggerSchema) validateField(value interface{}, fieldName, fieldType string, fieldDetails *swagger.ModelProperty) []error {
+	allErrs := []error{}
+	if reflect.TypeOf(value) == nil {
+		return append(allErrs, fmt.Errorf("unexpected nil value for field %v", fieldName))
+	}
 	// TODO: caesarxuchao: because we have multiple group/versions and objects
 	// may reference objects in other group, the commented out way of checking
 	// if a filedType is a type defined by us is outdated. We use a hacky way
@@ -310,7 +314,6 @@ func (s *SwaggerSchema) validateField(value interface{}, fieldName, fieldType st
 		// if strings.HasPrefix(fieldType, apiVersion) {
 		return s.ValidateObject(value, fieldName, fieldType)
 	}
-	allErrs := []error{}
 	switch fieldType {
 	case "string":
 		// Be loose about what we accept for 'string' since we use IntOrString in a couple of places

--- a/pkg/api/validation/schema_test.go
+++ b/pkg/api/validation/schema_test.go
@@ -209,6 +209,7 @@ func TestInvalid(t *testing.T) {
 		"invalidPod1.json", // command is a string, instead of []string.
 		"invalidPod2.json", // hostPort if of type string, instead of int.
 		"invalidPod3.json", // volumes is not an array of objects.
+		"invalidPod4.yaml", // string list with empty string.
 		"invalidPod.yaml",  // command is a string, instead of []string.
 	}
 	for _, test := range tests {

--- a/pkg/api/validation/testdata/v1/invalidPod4.yaml
+++ b/pkg/api/validation/testdata/v1/invalidPod4.yaml
@@ -1,0 +1,14 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    name: redis-master
+  name: name
+spec:
+  containers:
+  - image: gcr.io/fake_project/fake_image:fake_tag
+    name: master
+    args:
+    -
+    command:
+    -

